### PR TITLE
040 upgrade

### PR DESCRIPTION
**Pull request by DavidGoldwasser**
Upgrade newer 0.4.0 ext gem and Ruby 2.7

### CODE REVIEW

**Comment by DavidGoldwasser** [comment]
3 measures are failing

2 of the three commonly used create_bar measures. The DEER one is probably just not failing do to limited testing. The Unit tests for this method pass in th ext gem spec. Tests that are test_bad_fraction, test_bad_non_neg, test_bad_non_neg. More details are below.

`
Failure: test_bad_fraction(Minitest::Result) [/srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-extension-0.4.0/lib/openstudio/extension/core/os_lib_model_generation.rb:1733]: NoMethodError: undefined method `[]' for false:FalseClass /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-extension-0.4.0/lib/openstudio/extension/core/os_lib_model_generation.rb:1733:in `bar_from_building_type_ratios' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/lib/measures/create_bar_from_building_type_ratios/measure.rb:373:in `run' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/lib/measures/create_bar_from_building_type_ratios/tests/create_bar_from_building_type_ratios_test.rb:77:in `apply_measure_to_model' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/lib/measures/create_bar_from_building_type_ratios/tests/create_bar_from_building_type_ratios_test.rb:131:in `test_bad_fraction'
`

RadiantSlabWithDoas is failing with error for `NoMethodError: undefined method `to_CoilCoolingLowTempRadiantVarFlow' for #<OpenStudio::Model::OptionalHVACComponent:0x000055ff6ecd2968>`. More detail is below.

`
Failure: test_multi_zone_office_3C_ceiling(Minitest::Result) [/srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-standards-0.2.12/lib/openstudio-standards/prototypes/common/objects/Prototype.radiant_system_controls.rb:28]: NoMethodError: undefined method `to_CoilCoolingLowTempRadiantVarFlow' for #<OpenStudio::Model::OptionalHVACComponent:0x000055ff6ecd2968> /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-standards-0.2.12/lib/openstudio-standards/prototypes/common/objects/Prototype.radiant_system_controls.rb:28:in `model_add_radiant_proportional_controls' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-standards-0.2.12/lib/openstudio-standards/prototypes/common/objects/Prototype.hvac_systems.rb:4719:in `block in model_add_low_temp_radiant' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-standards-0.2.12/lib/openstudio-standards/prototypes/common/objects/Prototype.hvac_systems.rb:4638:in `each' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/.bundle/install/ruby/2.7.0/gems/openstudio-standards-0.2.12/lib/openstudio-standards/prototypes/common/objects/Prototype.hvac_systems.rb:4638:in `model_add_low_temp_radiant' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/lib/measures/radiant_slab_with_doas/measure.rb:312:in `run' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/lib/measures/radiant_slab_with_doas/tests/radiant_slab_with_doas_test.rb:109:in `apply_measure_and_run' /srv/jenkins/git/openstudio-model-articulation-gem/PR-61/1/lib/measures/radiant_slab_with_doas/tests/radiant_slab_with_doas_test.rb:296:in `test_multi_zone_office_3C_ceiling'
`

**Comment by DavidGoldwasser** [comment]
@mdahlhausen I'll look at the first one which is related to ext gem. Do you want to look at the second one which is related to standards gem. Not sure if the measure needs to be updated or change in standards gem.

**Comment by mdahlhausen** [comment]
looks like and SDK change going from HVACComponent to OptionalHVACComponent.

**Comment by mdahlhausen** [comment]
yep: https://github.com/NREL/OpenStudio/commit/0e6252d9a3f17c2c40400b15c1c276c247e899ee
And... I was the one who initially asked for this apparently. https://github.com/NREL/OpenStudio/issues/3532

I'll added a version check for this in the next version of openstudio-standards:
https://github.com/NREL/openstudio-standards/pull/1031

**Comment by DavidGoldwasser** [comment]
I just made a PR to extension gem that should fix the failure with invalid user argument. I added a test in extension gem for this. It broke when I refactored create bar methods and put double arg checks in dedicated method, and forgot to pass the return false back up to parent method.
https://github.com/NREL/openstudio-extension-gem/pull/114

**Comment by DavidGoldwasser** [comment]
@mdahlhausen I updated this PR to use master branch of standards, I see it is picking up your code and but because our version is still 3.1.1. and not 3.2 for now its going on the wrong side of the if statement. @tijcolem will next OS installer be a pre 3.2 or will it still be 3.1.1.

For reference it is failing on line 30 now. This is impacting PR on a least one other measure repo preventing upgrade to Ruby 2.7
https://github.com/NREL/openstudio-standards/commit/585adf71d284522684f25df289e6c88e9d74f61e

**Comment by DavidGoldwasser** [comment]
@mdahlhausen thanks for the update to standards, CI passed.

File: Gemfile [hunk 1]
```diff
@@ -11,6 +11,12 @@ gemspec
 # checkout the latest version (develop) from github.
 allow_local = ENV['FAVOR_LOCAL_GEMS']
 
+# uncomment when you want CI to use develop branch of extension gem
+gem 'openstudio-extension', github: 'NREL/OpenStudio-extension-gem', branch: 'develop'
```
Comment (kflemin) [inline Gemfile:15]: should these be uncommented?

**Review by kflemin — APPROVED** [review]
this seems fine as long as you don't make a new release in the 0.3.x range with these changes. but looks like you've already updated the version so👍

File: Gemfile (hunk 1 above)
Comment (DavidGoldwasser) [reply Gemfile:15]: Yes, they should stay un-commented for now, we could comment it out after the 3.2 release until we have fixes in extension gem or standards gem that are not in 3.2 OpenStudio installer.

**Pull request merged**